ow_power_system/src/PowerSystemNode.cpp

```diff
@@ -2,319 +2,374 @@
-
-      getline(line_stream, cell, ',');
-      Datum<double> power(stod(cell));
-      power.setTime(timestamp);
-
-      getline(line_stream, cell, ',');
-      Datum<double> temperature(stod(cell));
-      temperature.setTime(timestamp);
-
-      getline(line_stream, cell, ',');
-      Datum<double> voltage(stod(cell));
-      voltage.setTime(timestamp);
-
-      data.insert({ MessageId::Watts, power });
-      data.insert({ MessageId::Centigrade, temperature });
-      data.insert({ MessageId::Volts, voltage });
-
-      result.push_back(data);
-
-      line_number++;
+      prognosers.push_back(builder.build(m_power_models[i].bus,
+                                         m_power_models[i].name,
+                                         "trajectory"));
     }
   }
-  catch(...) // Many possible different errors could result from reading an improperly formatted file.
+  catch(const std::exception& err)
   {
-    ROS_ERROR_STREAM("Failed to read " << custom_file << ":" << endl <<
-                     "Improper formatting detected on line " << line_number << "." << endl << 
-                     "Confirm " << custom_file << " follows the exact format of example_fault.csv before retrying.");
-    return PrognoserVector();
+    ROS_ERROR_STREAM("OW_POWER_SYSTEM ERROR: " << err.what()
+                     << " while attempting to initialize asynchronous "
+                     << "prognosers!\nEnsure /config/async_prognoser.cfg exists "
+                     << "and is formatted properly before restarting the "
+                     << "simulation.");
+    return;
   }
-  return result;
-}
 
-bool PowerSystemNode::loadCustomFaultPowerProfile(string path, string custom_file)
-{
-  m_custom_power_fault_sequence = loadPowerProfile(path, custom_file);
+  ROS_INFO_STREAM("Power system node running.");
 
-  // Return false if the sequence was not properly initialized.
-  return (m_custom_power_fault_sequence.size() > 0);
-}
-
-bool PowerSystemNode::initPrognoser()
-{
-  // Create a configuration from a file
-  auto prognoser_config_path = ros::package::getPath("ow_power_system") + "/config/prognoser.cfg";
-  ConfigMap prognoser_config(prognoser_config_path);
+  // This rate object is used to sync the cycles up to the provided Hz.
+  // NOTE: If the cycle takes longer than the provided Hz, nothing bad will
+  //       necessarily occur, but the simulation will be out of sync with real
+  //       time.
+  ros::Rate rate(m_loop_rate_hz);
 
-  // Contruct a new prognoser using the prognoser factory. The prognoser
-  // will automatically construct an appropriate model, observer and predictor
-  // based on the values specified in the config.
-  m_prognoser = PrognoserFactory::instance().Create("ModelBasedPrognoser", prognoser_config);
+  // Start the asynchronous spinner, which will call jointStatesCb as the
+  // joint_states topic is published (50Hz).
+  ros::AsyncSpinner spinner(m_spinner_threads);
+  spinner.start();
```

Comment:
I have to declare ``spinner`` outside of a potential try-catch statement due to scope, so if ``spinner.start()`` is the specific line that can throw the error, it's fine. If not, then I might have to encase most of the loop in a try-catch statement which probably isn't ideal.

Mostly because I want to learn, how can this throw a ``std::runtime_error``? I'm not too well-versed in multithreading to begin with.

Reply:
Well written documentation will tell you exactly what exceptions a function may throw, but this is ROS...

Honestly it's pure serendipity that I even noticed it may throw a `std::runtime_error` because I happened to take a look at its [source code](http://docs.ros.org/en/noetic/api/roscpp/html/spinner_8cpp_source.html), but I cannot recall why. 

Shifting through source code is not my typical practice when I want to learn about an API, but it is one way to work around lacking documentation.

Reply:
Sorry I highlighted more code than needed, but I meant that only `start` may throw.